cpp/src/barretenberg/proof_system/composer/permutation_helper.hpp

```diff
@@ -442,4 +443,97 @@ inline void compute_first_and_last_lagrange_polynomials(auto key) // proving_key
     key->polynomial_store.put("L_last_lagrange", std::move(lagrange_polynomial_n_min_1));
 }
 
+/**
+ * @brief Compute generalized permutation sigmas and ids for ultra plonk
+ *
+ * @tparam program_width
+ * @tparam CircuitConstructor
+ * @param circuit_constructor
+ * @param key
+ * @return std::array<std::vector<permutation_subgroup_element>, program_width>
+ */
+template <size_t program_width, typename CircuitConstructor>
+void compute_plonk_generalized_sigma_permutations(const CircuitConstructor& circuit_constructor, bonk::proving_key* key)
+{
+    // Compute wire copy cycles for public and private variables
+    auto wire_copy_cycles = compute_wire_copy_cycles<program_width>(circuit_constructor);
+    std::array<std::vector<permutation_subgroup_element>, program_width> sigma_mappings;
+    std::array<std::vector<permutation_subgroup_element>, program_width> id_mappings;
+
+    // Instantiate the sigma and id mappings by reserving enough space and pushing 'default' permutation subgroup
+    // elements that point to themselves.
+    for (size_t i = 0; i < program_width; ++i) {
+        sigma_mappings[i].reserve(key->circuit_size);
+        id_mappings[i].reserve(key->circuit_size);
```

Comment:
what happened to `with_tags`?

Reply:
This used to distinguish between generalized vs non-generalized permutations? I suppose we'll leave it up to @Rumata888 about whether to combine these functions again or not?

Reply:
Yeah the switch was redundant since this function is only used for ultra. But could be reintroduced if it makes sense to @Rumata888 to combine things.